File: optimised_repeatingCharacter.java
import java.util.HashMap;

/**
 * optimised_repeatingCharacter
 */
public class optimised_repeatingCharacter {

    public static int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> mp = new HashMap<Character, Integer>();
        int left = 0, right = 0;
        int n = s.length();
        int len = 0;
        while (right < n) {
            if (mp.containsKey(s.charAt(right))) {
                left = Math.max(mp.get(s.charAt(right)) + 1, left);
            }
            mp.put(s.charAt(right), right);
            len = Math.max(len, right - left + 1);
            right++;
        }
        return len;
    }

    public static void main(String[] args) {
        String str = "takeUforward";
        System.out.println("The longest subsequence is " + lengthOfLongestSubstring(str));
    }
}

File: optimised_repeatingCharacter.py
def lengthOfLongestSubstring(s: str) -> int:
    mp = [-1] * 256
    
    left, right = 0, 0
    n = len(s)
    length = 0
    while right < n:
        if mp[ord(s[right])] != -1:
            left = max(mp[ord(s[right])] + 1, left)
        
        mp[ord(s[right])] = right
        length = max(length, right - left + 1)
        right += 1
    return length

if __name__ == "__main__":
    a = "takeUforward"
    print("The longest subsequence is ", lengthOfLongestSubstring(a))
    